File: leetcode_other/Arrays and Strings/maximum_number_of_balloons.py
class Solution:
    def maxNumberOfBalloons(self, text: str) -> int:
        """
        Concept:
        Hashmaps
        Create the lookup dicts for balloon, and for the text string,
        only get the characters that exist in the word "balloon"

        if the length of the keys in lookup and textLookup aren't the same,
        that means we don't have all the characters necessary to form the word "balloon", so return 0

        repeatedly subtract the count from each key in lookup dict, if all values are >= 0,
        that means we have successfully formed the word "balloon", so increment by 1

        If at any point, the count of any characters necessary drops below 0, that means we didn't fully
        form the word balloon, so we break and return the count

        """
        from collections import Counter
        lookup = Counter("balloon")
        textLookup = dict()
        for i in range(len(text)):
            if text[i] in lookup:
                if text[i] in textLookup:
                    textLookup[text[i]] += 1
                else:
                    textLookup[text[i]] = 1
        if len(lookup.keys()) != len(textLookup.keys()):
            return 0
        
        count = 0
        while (True):
            textLookup["b"] -= 1
            textLookup["a"] -= 1
            textLookup["l"] -= 2
            textLookup["o"] -= 2
            textLookup["n"] -= 1
            if (textLookup["b"] >= 0 and textLookup["a"] >= 0 
                and textLookup["l"] >= 0 and textLookup["o"] >= 0
                and textLookup["n"] >= 0
               ):
                count += 1
            else:
                break
        return count
        

File: node_js_ds_and_algos_implementations/arrays and strings/hashmaps/maximum_number_of_balloons.js
/**
 * @param {string} text
 * @return {number}
 */
/*
Approach:

get counts of all chars in text
s = "balloon"
total = 0 
while (true)
    found = 0
    for each char in s
        if char in counter and counter[char] > 0
            decrement counter[char]
            increment found
        else
            break, because we can't form the word balloon if we can't find all chars
    // if we've found all the characters for "balloon", increment total by 1
    if found === s.length
        increment total
    else
        break
return total

O(N^2) time
O(N) space

*/
var maxNumberOfBalloons = function(text) {
    let counter = {}
    for (let c of text){
        if (c in counter){
            ++counter[c]
        }
        else {
            counter[c] = 1
        }
    }
    let count = 0
    let s = "balloon"

    while (true){
        let found = 0
        for (let c of s){
            if (c in counter && counter[c] > 0){
                counter[c]--
                ++found
            }
            else {
                break
            }
        }
        if (found === s.length){
            ++count
        }
        else {
            break
        }
    }
    return count
};